### PR TITLE
Prevent sticky & auto-assign of ambiguous devices

### PR DESCRIPTION
**Pull request by pearsonk**
Add checks for device ambiguity to:
- Setting the sticky but
- Automatic assignment of new device
- Automatic assignment of created VM

This prevents policy rules from matching multiple devices, each of which
would be passed to a single VM instance. This is most common when
devices do not populate the serial number, thus there is not enough
information to uniquely identify based on policy rules. This does not
affect manual assignment, which uses USB bus and device IDs as opposed
to the vendor/product IDs of common rules such as the sticky/ALWAYS
mechanism.

OXT-1782

Signed-off-by: Kevin Pearson <kevin.pearson@ortmanconsulting.com>

### CODE REVIEW

File: src/policy.c [hunk 1]
```diff
@@ -348,6 +354,15 @@ policy_auto_assign_new_device(device_t *device)
   int uivm;
   int res = 1;
 
+  if (device == NULL) return 1;
+
+  /* Don't auto assign ambiguous devices */
+  if (device_is_ambiguous(device))
+  {
+    xd_log(LOG_INFO, "Rejecting automatic assignment of ambiguous device: Bus=%d Dev=%d", device->busid, device->devid);
```
Comment (jandryuk) [inline src/policy.c]: This line is kinda long, you can leave the string together but break the arguments up over multiple lines?

File: src/policy.c [hunk 2]
```diff
@@ -229,6 +229,12 @@ policy_set_sticky(int dev)
   device = device_lookup(busid, devid);
   if (device == NULL || device->vm == NULL)
     return -1;
+  /* Do not set sticky for ambiguous devices */
+  if (device_is_ambiguous(device) == 1)
```
Comment (jandryuk) [inline src/policy.c]: Just `if (device_is_ambiguous(device))`?

File: src/policy.c [hunk 3]
```diff
@@ -388,14 +403,15 @@ policy_auto_assign_new_device(device_t *device)
...
+  struct list_head *pos, *device_pos, *tmp;
   rule_t *rule;
   device_t *device;
   int ret = 0;
+  int clean = 0;
 
   /* For all the ALWAYS and DEFAULT rules that match the VM,
    * assign all devices that match the rule to the VM */
-  list_for_each(pos, &rules.list) {
+  list_for_each_safe(pos, tmp, &rules.list) {
```
Comment (jandryuk) [inline src/policy.c:423]: Is converting to `list_for_each_safe` necessary or an unrelated change?

File: src/policy.c [hunk 4]
```diff
@@ -416,13 +432,33 @@ policy_auto_assign_devices_to_new_vm(vm_t *vm)
...
           device->vm = vm;
           ret |= -usbowls_plug_device(vm->domid, device->busid, device->devid);
         }
       }
     }
+    /* Cleanse rule because UI thinks unassigned devices are attached due to
+     * sticky association */
+    if (clean == 1){
+      list_del(&rule->list);
+      dump_rules();
```
Comment (jandryuk) [inline src/policy.c]: Do we really want to dump the whole rules DB?   Maybe just print that we are deleting the rule, maybe with some details?

**Review by jandryuk — COMMENTED** [review]
Thanks Kevin, this generally looks good.  I have a few formatting suggestions to keep the lines under 80 characters, but there are a few more I didn't mark.  Looks like the source doesn't always follow that, but I'd prefer to keep lines shorter if possible.

File: src/policy.c (hunk 2 above)
Comment (pearsonk) [reply src/policy.c]: Since it's an int and I differentiated between error (-1), non-ambiguous (0), and ambiguous (1), I made this check explicit to not proceed on error.

File: src/policy.c (hunk 3 above)
Comment (pearsonk) [reply src/policy.c:423]: This is necessary for continued safe traversal because of the delete added later. Otherwise it segfaults due to the list_head manipulation.

File: src/policy.c (hunk 4 above)
Comment (pearsonk) [reply src/policy.c]: oversight, will remove the dump

File: src/policy.c (hunk 1 above)
Comment (pearsonk) [reply src/policy.c]: ack, will do some format cleanup

File: src/policy.c (hunk 2 above)
Comment (jandryuk) [reply src/policy.c]: Ah, okay.  Hmmm, you have two other instances with just `if (device_is_ambiguous(device))`.  `device_is_ambiguous()` only returns -1 when the pass-ed in device is NULL, and you have a bunch of those NULL checks already and even add some.  Maybe just make `device_is_ambiguous()` return only 0 and 1 and avoid the -1 case?

Comment (pearsonk) [reply src/policy.c]: Makes sense, adjusted the function in device.c to only return 0 or 1 to treat as a cleaner boolean

**Comment by pearsonk** [comment]
Added a couple commits that will be squashed later, but have observed force-push rebases killing cited questions/comments in the past. I'll squash to a single commit once the initial source review is done.

File: src/policy.c [hunk 5]
```diff
@@ -214,7 +214,7 @@ default_lookup(device_t *device)
  *
  * @param dev The device single ID
  *
- * @return 0 if the device was found and assigned to a VM, -1 otherwise
+ * @return 0 if the device was found and assigned to a VM, 1 if the device is ambiguous, -1 otherwise
```
Comment (jandryuk) [inline src/policy.c]: This line is still long.

**Review by jandryuk — COMMENTED** [review]
Looks like your rebase was a little wonky since it grabbed earlier commits.

File: src/policy.c (hunk 5 above)
Comment (pearsonk) [reply src/policy.c]: Updated, and fixed the wonky rebase

**Comment by jandryuk** [comment]
Thanks.  I'll run a test build locally.

**Comment by eric-ch** [comment]
Build [10099](http://openxt-builder.ainfosec.com:8010/#/builders/3/builds/10099)

**Comment by jandryuk** [comment]
I'm planning to merge this later today.

I don't actually have identical devices to test the ambiguity check.  It looks okay from code review and doesn't seem to break regular assignment.
